cpp/CMakeLists.txt

```diff
@@ -783,24 +543,133 @@ set_property(CACHE RMM_LOGGING_LEVEL PROPERTY STRINGS
         "TRACE" "DEBUG" "INFO" "WARN" "ERROR" "CRITICAL" "OFF")
 message(STATUS "RMM_LOGGING_LEVEL = '${RMM_LOGGING_LEVEL}'.")
 
-target_compile_definitions(cudf PUBLIC SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_${RMM_LOGGING_LEVEL})
+###################################################################################################
+# - library targets -------------------------------------------------------------------------------
+
+function(add_library_module NAMESPACE MODULE MODULE_DIR)
+    file(GLOB_RECURSE MODULE_SRC "${MODULE_DIR}/*")
+    add_library("${NAMESPACE}_${MODULE}" ${MODULE_SRC})
+    set_target_properties("${NAMESPACE}_${MODULE}" PROPERTIES BUILD_RPATH "\$ORIGIN")
+    add_dependencies("${NAMESPACE}_${MODULE}" stringify_run)
+    target_compile_definitions("${NAMESPACE}_${MODULE}" PUBLIC "SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_${LOGGING_LEVEL}")
+    add_library("${NAMESPACE}::${MODULE}" ALIAS "${NAMESPACE}_${MODULE}")
+endfunction()
+
+set(CUDF_NAMESPACE "cudf")
+set(CUDF_SRC "${CMAKE_BINARY_DIR}/cudf.cu")  # this file will be generated
+
+set(
+    CUDF_HEADERS_DIRS
+        "${CMAKE_SOURCE_DIR}/include/cudf"
+        "${CMAKE_SOURCE_DIR}/include/nvtext"
+)
+
+set(
+    CUDF_BASE_MODULE_LIST
+        "${CMAKE_SOURCE_DIR}/src/aggregation"
+        "${CMAKE_SOURCE_DIR}/src/binaryop"
+        "${CMAKE_SOURCE_DIR}/src/bitmask"
+        "${CMAKE_SOURCE_DIR}/src/column"
+        "${CMAKE_SOURCE_DIR}/src/copying"
+        "${CMAKE_SOURCE_DIR}/src/datetime"
+        "${CMAKE_SOURCE_DIR}/src/dictionary"
+        "${CMAKE_SOURCE_DIR}/src/filling"
+        "${CMAKE_SOURCE_DIR}/src/groupby"
+        "${CMAKE_SOURCE_DIR}/src/jit"
+        "${CMAKE_SOURCE_DIR}/src/lists"
+        "${CMAKE_SOURCE_DIR}/src/quantiles"
+        "${CMAKE_SOURCE_DIR}/src/reshape"
+        "${CMAKE_SOURCE_DIR}/src/scalar"
+        "${CMAKE_SOURCE_DIR}/src/search"
+        "${CMAKE_SOURCE_DIR}/src/sort"
+        "${CMAKE_SOURCE_DIR}/src/stream_compaction"
+        "${CMAKE_SOURCE_DIR}/src/strings"
+        "${CMAKE_SOURCE_DIR}/src/structs"
+        "${CMAKE_SOURCE_DIR}/src/table"
+        "${CMAKE_SOURCE_DIR}/src/text"
+        "${CMAKE_SOURCE_DIR}/src/transform"
+        "${CMAKE_SOURCE_DIR}/src/unary"
+        "${CMAKE_SOURCE_DIR}/src/utilities"
+)
```

Comment:
No issues, just a comment that this base list is huge and we could likely do with some refactoring to better split this up.

Reply:
I agree. More comments would be helpful. Unfortunately, I don't think splitting this up will be possible. These modules are grouped together because they have cyclic dependencies (there's not a nice partitioning).

As an example: `dictionary` calls `encode` from `transform`. Pulling in `transform` pulls in every symbol/declaration that is not encapsulated by `transform`.

Reply:
Yup, my point is we should keep this in mind as we continue to develop libcudf to resolve / avoid these types of cyclic dependencies if possible.